db/db_impl.cc

```diff
@@ -1639,6 +1641,271 @@ std::vector<Status> DBImpl::MultiGet(
   return stat_list;
 }
 
+// Order keys by CF ID, followed by key contents
+inline bool CompareKeyContext(const KeyContext* lhs, const KeyContext* rhs) {
+  auto cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(lhs->column_family);
+  auto cfd = cfh->cfd();
+  auto cf_id1 = cfd->GetID();
+  cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(rhs->column_family);
+  cfd = cfh->cfd();
+  auto cf_id2 = cfd->GetID();
+  if (cf_id1 < cf_id2) {
+    return true;
+  } else if (cf_id1 > cf_id2) {
+    return false;
+  }
+
+  auto comparator = cfd->user_comparator();
+  int cmp = comparator->Compare(*(lhs->key), *(rhs->key));
+  if (cmp < 0) {
+    return true;
+  }
+  return false;
+}
+
+void DBImpl::MultiGet(const ReadOptions& read_options,
+                      const std::vector<ColumnFamilyHandle*>& column_family,
+                      const std::vector<Slice>& keys, PinnableSlice* values,
+                      Status* statuses) {
+  autovector<KeyContext, MultiGetContext::MAX_KEYS_ON_STACK> key_context;
+  size_t index = 0;
+  for (const Slice& user_key : keys) {
+    key_context.emplace_back(user_key, column_family[index], &values[index],
+                             &statuses[index]);
+    index++;
+  }
+
+  MultiGetImpl(read_options, key_context);
+}
+
+void DBImpl::MultiGetImpl(const ReadOptions& read_options,
+      autovector<KeyContext, MultiGetContext::MAX_KEYS_ON_STACK>& key_context,
+      ReadCallback* callback, bool* is_blob_index) {
+  PERF_CPU_TIMER_GUARD(get_cpu_nanos, env_);
+  StopWatch sw(env_, stats_, DB_MULTIGET);
+  size_t num_keys = key_context.size();
+
+  PERF_TIMER_GUARD(get_snapshot_time);
+
+  SequenceNumber snapshot;
+
+  struct MultiGetColumnFamilyData {
+    ColumnFamilyData* cfd;
+    size_t start;
+    size_t end;
+    SuperVersion* super_version;
+    MultiGetColumnFamilyData(ColumnFamilyData* cf, size_t first, size_t last,
+                             SuperVersion* sv)
+        : cfd(cf), start(first), end(last), super_version(sv) {}
+    MultiGetColumnFamilyData() = default;
+  };
+
+  autovector<MultiGetColumnFamilyData, MultiGetContext::MAX_KEYS_ON_STACK>
+      multiget_cf_data;
+  autovector<KeyContext*, MultiGetContext::MAX_KEYS_ON_STACK> sorted_keys;
+  sorted_keys.resize(num_keys);
+  {
+    size_t index = 0;
+    for (KeyContext& key : key_context) {
+      sorted_keys[index] = &key;
+      index++;
+    }
+    std::sort(sorted_keys.begin(), sorted_keys.begin() + index,
+              CompareKeyContext);
```

Comment:
I wonder whether sometimes users have pre-sorted the keys so that we don't need to do it here. If it is possible, we should have an argument to skip this step. Maybe verify it in DEBUG mode.

Reply:
I'll add an argument to skip the sorting